### ISSUE TITLE
Download exhentai galleries from mirror sites (hitomi.la and hiyobi.me)

### ISSUE THREAD

**Issue by lukemin** [HDoujinDownloader/HDoujin-Downloader#8]
Can Hdoujin-downloader download mangas of exhentai from hitomi.la and hiyobi.me instead？ That can speedup download speed and save lots of GP of exhentai.

**Comment by HDoujinDownloader** [comment]
That's definitely a cool idea. I think hitomi.la at least has a one-to-one mapping of exhentai gallery titles to their own, so I don't think it should be too difficult to add.

**Comment by lukemin** [comment]
If you can add this function, please forget my "Ask for refund" letter, I updated Hdoujin Downloader to Full version 2 days before. But unfortunately I found that I did not have enough gp to download galleries from exhentai. One more thing, download speed of exhentai is about 20 times slower than from nhentai and hitomi. When I set max simultaneous downloads to 20，I got banned by exhentai for more than 2 days.

**Comment by HDoujinDownloader** [comment]
I've got this working now on my end, and it's now able to find mirrors on hitomi.la and hiyobi.me for exhentai.org galleries (as well as e-hentai.org galleries). I'll probably upload an update sometime tomorrow.

Aside from speeding up downloads and avoiding spending GP, another nice benefit of this is that it also eliminates the need to provide a login when downloading from exhentai.org when a mirror can be found.

>  When I set max simultaneous downloads to 20，I got banned by exhentai for more than 2 days.

Yeah, they're pretty hostile to automated scraping (I can't really blame them, though). In fact, the ability to set timing settings under `Tools > Settings > Downloads > Timing` was added specifically to help avoid this situation. It'll be less of an issue with this next update, but it might be worth being aware of nonetheless.

**Comment by lukemin** [comment]
nice to hear that， thanks a lot.
I have another useful feture request. Can I post it in this issue？ or just start an new one？

**Comment by HDoujinDownloader** [comment]
Unless it's directly related to this one, it's probably best to start a new one.

**Comment by HDoujinDownloader** [comment]
All right, it's now possible to download from mirror sites as of `2020.12.24.0-alpha`, which you can download from the [Releases page](https://github.com/HDoujinDownloader/HDoujin-Downloader/releases/tag/v2020.12.24.0-alpha) (just extract it over your current HDoujin Downloader folder to "update"). This feature is not enabled by default, and needs to be explicitly toggled under `Tools > Site preferences > e-hentai.org > Images > Download from mirror sites`.

I'm not releasing this as an official update just yet (I'm still in the middle of other changes I'm working on at the same time), but it should be stable.

Let me know if you have any problems!

**Comment by lukemin** [comment]
when I extracted 2020.12.24.0-alpha to the current HDoujin Downloader folder, all the configs were back to default. 
The hello popup window came out everytime evenif **donot show** opinion was selected.
![2020-12-25_203028](https://user-images.githubusercontent.com/12003375/103135024-5a89ae80-46f0-11eb-96fb-d574d9778137.png)

The proxy setting doesn't show proxy type and could not connect to my local proxy server.
![2020-12-25_202810](https://user-images.githubusercontent.com/12003375/103135027-5eb5cc00-46f0-11eb-8e6f-8a06dd15c956.png)

**Comment by HDoujinDownloader** [comment]
I wonder if it's antivirus software specifically blocking access to the `config.ini` file because it doesn't recognize the new executable (as INI files are considered "system files" on Windows). The read indicated in the log is probably failing because the write failed. Try closing HDoujin Downloader and then looking at the latest item in the `log\` directory for any errors occurring while writing settings (such as an `UnauthorizedAccessException`).

You can also restore your previous `config.ini` from one of the backup archives in the `backup/` directory (they're just ZIP archives).

Some things to try:
* Whitelist HDoujin Downloader in your antivirus software
* Try extracting the updated version to a new directory
* Make sure nothing is locking the `config.ini` file

**Comment by lukemin** [comment]
I extracted the updated version to a new directory，same problem.

**Comment by HDoujinDownloader** [comment]
Did you try whitelisting it in your antivirus software? What does the log file say when you close the application? I mention closing it specifically because that's when the `config.ini` file is updated, so if that's the part that's failing, it would be logged at that moment.

**Comment by lukemin** [comment]
Uploading bandicam 2020-12-25 21-30-11-339.mp4…

**Comment by lukemin** [comment]
https://user-images.githubusercontent.com/12003375/103136137-bf490700-46f8-11eb-9a2d-6317bfbac9c6.mp4

**Comment by HDoujinDownloader** [comment]
You went through the correct steps! When you toggle `Help > Enable Diagnostics`, it starts writing a log file in the `log/` directory. If you open it, you'll see files with names like `<timestamp>-debug.log`. Open the one with the largest timestamp in Notepad, and look for errors relating to writing the settings file. Alternatively, you can just email me the log file and I'll take a look at it.

**Comment by lukemin** [comment]
all settings cannot be saved including the setting for Diagnostics.
Run as administrator does not work
[1608904400-debug.log](https://github.com/HDoujinDownloader/HDoujin-Downloader/files/5742036/1608904400-debug.log)

**Comment by HDoujinDownloader** [comment]
Thanks a lot for including the log file. Unfortunately, I don't see anything related to the issue you're currently experiencing. However, I think I know what the problem is. Give me just a moment, and I'll upload a fix.

**Comment by HDoujinDownloader** [comment]
All right, thank you for your patience! I've uploaded a fix for what I believe was the problem-- Download it [here](https://github.com/HDoujinDownloader/HDoujin-Downloader/releases/tag/v2020.12.24.0-alpha).

**Comment by lukemin** [comment]
After testing，the function (Download exhentai galleries from mirror sites ) does not work. 
eg:
https://exhentai.org/g/1078396/ef2d7ca909/
https://hitomi.la/manga/ninshin-paradox-%E4%B8%AD%E6%96%87-1078396.html

When closing app with downloading tasks, the app crashed.


有关调用实时(JIT)调试而不是此对话框的详细信息，
请参见此消息的结尾。

************** 异常文本 **************
System.InvalidOperationException: 集合已修改；可能无法执行枚举操作。
   在 System.Collections.ArrayList.ArrayListEnumeratorSimple.MoveNext()
   在 System.Windows.Forms.Application.ExitInternal()
   在 System.Windows.Forms.Application.Exit(CancelEventArgs e)
   在 \*K<\&\&=54uj|=)uJWz)LD`sM;%.‮‏‎⁪⁮⁪‭‍‮‫​​‎‍‮‭⁭‬⁪⁪‮‬⁪​‮⁯‬‍‍‬‮⁫‪⁯⁪‫⁫​‍⁯‮(Object , EventArgs )
   在 System.Windows.Forms.Timer.OnTick(EventArgs e)
   在 System.Windows.Forms.Timer.TimerNativeWindow.WndProc(Message& m)
   在 System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, Int32 msg, IntPtr wparam, IntPtr lparam)


************** 已加载的程序集 **************
mscorlib
    程序集版本:4.0.0.0
    Win32 版本:4.8.4250.0 built by: NET48REL1LAST_C
    基本代码:file:///C:/Windows/Microsoft.NET/Framework64/v4.0.30319/mscorlib.dll
----------------------------------------
HDoujin Downloader
    程序集版本:2020.12.24.3
    Win32 版本:2020.12.24.3
    基本代码:file:///D:/HDoujin.Downloader-2020.8.8.0-win/HDoujin%20Downloader.exe
----------------------------------------
ZQZHZayBIVKvCEqoekcyLjuTNRbO
    程序集版本:0.0.0.0
    Win32 版本:2020.12.24.3
    基本代码:file:///D:/HDoujin.Downloader-2020.8.8.0-win/HDoujin%20Downloader.exe
----------------------------------------
Microsoft.VisualBasic
    程序集版本:10.0.0.0
    Win32 版本:14.8.3752.0 built by: NET48REL1
    基本代码:file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/Microsoft.VisualBasic/v4.0_10.0.0.0__b03f5f7f11d50a3a/Microsoft.VisualBasic.dll
----------------------------------------
System
    程序集版本:4.0.0.0
    Win32 版本:4.8.4200.0 built by: NET48REL1LAST_C
    基本代码:file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System/v4.0_4.0.0.0__b77a5c561934e089/System.dll
----------------------------------------
System.Core
    程序集版本:4.0.0.0
    Win32 版本:4.8.4270.0 built by: NET48REL1LAST_C
    基本代码:file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.Core/v4.0_4.0.0.0__b77a5c561934e089/System.Core.dll
----------------------------------------
System.Windows.Forms
    程序集版本:4.0.0.0
    Win32 版本:4.8.4270.0 built by: NET48REL1LAST_C
    基本代码:file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.Windows.Forms/v4.0_4.0.0.0__b77a5c561934e089/System.Windows.Forms.dll
----------------------------------------
System.Drawing
    程序集版本:4.0.0.0
    Win32 版本:4.8.3752.0 built by: NET48REL1
    基本代码:file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.Drawing/v4.0_4.0.0.0__b03f5f7f11d50a3a/System.Drawing.dll
----------------------------------------
System.Configuration
    程序集版本:4.0.0.0
    Win32 版本:4.8.4190.0 built by: NET48REL1LAST_B
    基本代码:file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.Configuration/v4.0_4.0.0.0__b03f5f7f11d50a3a/System.Configuration.dll
----------------------------------------
System.Xml
    程序集版本:4.0.0.0
    Win32 版本:4.8.3752.0 built by: NET48REL1
    基本代码:file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.Xml/v4.0_4.0.0.0__b77a5c561934e089/System.Xml.dll
----------------------------------------
System.Runtime.Remoting
    程序集版本:4.0.0.0
    Win32 版本:4.8.3752.0 built by: NET48REL1
    基本代码:file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.Runtime.Remoting/v4.0_4.0.0.0__b77a5c561934e089/System.Runtime.Remoting.dll
----------------------------------------
System.Windows.Forms.resources
    程序集版本:4.0.0.0
    Win32 版本:4.8.3752.0 built by: NET48REL1
    基本代码:file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.Windows.Forms.resources/v4.0_4.0.0.0_zh-Hans_b77a5c561934e089/System.Windows.Forms.resources.dll
----------------------------------------
System.Numerics
    程序集版本:4.0.0.0
    Win32 版本:4.8.3752.0 built by: NET48REL1
    基本代码:file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.Numerics/v4.0_4.0.0.0__b77a5c561934e089/System.Numerics.dll
----------------------------------------
System.Runtime.Serialization
    程序集版本:4.0.0.0
    Win32 版本:4.8.4250.0 built by: NET48REL1LAST_C
    基本代码:file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.Runtime.Serialization/v4.0_4.0.0.0__b77a5c561934e089/System.Runtime.Serialization.dll
----------------------------------------
System.Data
    程序集版本:4.0.0.0
    Win32 版本:4.8.4270.0 built by: NET48REL1LAST_C
    基本代码:file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_64/System.Data/v4.0_4.0.0.0__b77a5c561934e089/System.Data.dll
----------------------------------------
MoonSharp.Interpreter
    程序集版本:2.0.0.0
    Win32 版本:2.0.0.0
    基本代码:file:///D:/HDoujin.Downloader-2020.8.8.0-win/lib/MoonSharp.Interpreter.dll
----------------------------------------
System.Security
    程序集版本:4.0.0.0
    Win32 版本:4.8.4261.0 built by: NET48REL1LAST_B
    基本代码:file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.Security/v4.0_4.0.0.0__b03f5f7f11d50a3a/System.Security.dll
----------------------------------------
Accessibility
    程序集版本:4.0.0.0
    Win32 版本:4.8.3752.0 built by: NET48REL1
    基本代码:file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/Accessibility/v4.0_4.0.0.0__b03f5f7f11d50a3a/Accessibility.dll
----------------------------------------
System.Data.SqlXml
    程序集版本:4.0.0.0
    Win32 版本:4.8.3752.0 built by: NET48REL1
    基本代码:file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.Data.SqlXml/v4.0_4.0.0.0__b77a5c561934e089/System.Data.SqlXml.dll
----------------------------------------
System.Web
    程序集版本:4.0.0.0
    Win32 版本:4.8.4250.0 built by: NET48REL1LAST_C
    基本代码:file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_64/System.Web/v4.0_4.0.0.0__b03f5f7f11d50a3a/System.Web.dll
----------------------------------------
System.ServiceModel.Internals
    程序集版本:4.0.0.0
    Win32 版本:4.8.4250.0 built by: NET48REL1LAST_C
    基本代码:file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.ServiceModel.Internals/v4.0_4.0.0.0__31bf3856ad364e35/System.ServiceModel.Internals.dll
----------------------------------------
System.Transactions
    程序集版本:4.0.0.0
    Win32 版本:4.8.3752.0 built by: NET48REL1
    基本代码:file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_64/System.Transactions/v4.0_4.0.0.0__b77a5c561934e089/System.Transactions.dll
----------------------------------------
Anonymously Hosted DynamicMethods Assembly
    程序集版本:0.0.0.0
    Win32 版本:4.8.4250.0 built by: NET48REL1LAST_C
    基本代码:file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_64/mscorlib/v4.0_4.0.0.0__b77a5c561934e089/mscorlib.dll
----------------------------------------
WebDriver
    程序集版本:3.141.0.0
    Win32 版本:3.141.0.0
    基本代码:file:///D:/HDoujin.Downloader-2020.8.8.0-win/lib/WebDriver.dll
----------------------------------------
Jint
    程序集版本:0.0.0.0
    Win32 版本:2.11.58.0
    基本代码:file:///D:/HDoujin.Downloader-2020.8.8.0-win/lib/Jint.dll
----------------------------------------
mscorlib.resources
    程序集版本:4.0.0.0
    Win32 版本:4.8.3752.0 built by: NET48REL1
    基本代码:file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/mscorlib.resources/v4.0_4.0.0.0_zh-Hans_b77a5c561934e089/mscorlib.resources.dll
----------------------------------------
System.ComponentModel.DataAnnotations
    程序集版本:4.0.0.0
    Win32 版本:4.8.3752.0
    基本代码:file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.ComponentModel.DataAnnotations/v4.0_4.0.0.0__31bf3856ad364e35/System.ComponentModel.DataAnnotations.dll
----------------------------------------
System.Web.ApplicationServices
    程序集版本:4.0.0.0
    Win32 版本:4.8.4250.0
    基本代码:file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.Web.ApplicationServices/v4.0_4.0.0.0__31bf3856ad364e35/System.Web.ApplicationServices.dll
----------------------------------------
System.Runtime.Caching
    程序集版本:4.0.0.0
    Win32 版本:4.8.3752.0
    基本代码:file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.Runtime.Caching/v4.0_4.0.0.0__b03f5f7f11d50a3a/System.Runtime.Caching.dll
----------------------------------------
Microsoft.Build.Framework
    程序集版本:4.0.0.0
    Win32 版本:4.8.3752.0 built by: NET48REL1
    基本代码:file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/Microsoft.Build.Framework/v4.0_4.0.0.0__b03f5f7f11d50a3a/Microsoft.Build.Framework.dll
----------------------------------------
System.resources
    程序集版本:4.0.0.0
    Win32 版本:4.8.3752.0 built by: NET48REL1
    基本代码:file:///C:/WINDOWS/Microsoft.Net/assembly/GAC_MSIL/System.resources/v4.0_4.0.0.0_zh-Hans_b77a5c561934e089/System.resources.dll
----------------------------------------

************** JIT 调试 **************
要启用实时(JIT)调试，
该应用程序或计算机的 .config 文件(machine.config)的 system.windows.forms 节中必须设置
jitDebugging 值。
编译应用程序时还必须启用
调试。

例如:

<configuration>
    <system.windows.forms jitDebugging="true" />
</configuration>

启用 JIT 调试后，任何未经处理的异常
都将被发送到在此计算机上注册的 JIT 调试程序，
而不是由此对话框处理。

**Comment by HDoujinDownloader** [comment]
I just uploaded version [`2020.12.24.4-alpha`](https://github.com/HDoujinDownloader/HDoujin-Downloader/releases/tag/v2020.12.24.0-alpha), which should fix the crash on exit. Let me know how it goes for you.

> the function (Download exhentai galleries from mirror sites ) worked partly

Is the example gallery you provided not working for you? It might be worth noting that the process is transparent, and as long as the `Download from mirror sites` setting is checked, you should be able to see it accessing the mirrors in the log (`Tools > Log Viewer`):

```
06:25:37    hdoujin down[i] Downloading from mirror: https://hitomi.la/galleries/1078396.html
06:25:37    download    [i] Downloading https://bb.hitomi.la/images/4/70/8f229d76b389192907a3eb622266953f28930453b433118259c877eb4caa8704.jpg
06:25:37    download    [i] Downloading https://bb.hitomi.la/images/f/8e/783299071b753e71f4d9c526a6284bcafcc62ecde8bc529e8a6d70afd2bc28ef.jpg
```
Before downloading from a mirror, it attempts to locate the gallery on one of the mirror websites (either `hitomi.la` or `hiyobi.me`). If the site can't be contacted for whatever reason, it will not use the mirror.

**Comment by lukemin** [comment]
The connection between me and hitomi.la is a bit bad, sometimes failed. maybe this is the reason.

**Comment by HDoujinDownloader** [comment]
It could definitely cause problems if there are intermittent connection failures. This could be handled either by aborting the download if a mirror can't be found, or retrying a number of times. If you have an opinion on this, let me know.

**Comment by lukemin** [comment]
Retrying after setting time, or skipping the failures to start manually maybe good solution.
-----
BUT I meet another confusing thing - The speed of downloading manges directle from hitomi.la(max speed upto 10MB/s) is about 10 times than downloading hitomi.la as a mirror of exhentai.org(max speed only 1MB/s).

**Comment by HDoujinDownloader** [comment]
> Retrying after setting time, or skipping the failures to start manually maybe good solution.

I've now added both of these options, so expect them to be present when I upload the next update. They'll be under a new section, `Tools > Preferences > Advanced`, because I consider them to be niche settings that most users won't need.

> The speed of downloading manges directly from hitomi.la(max speed upto 10MB/s) is about 10 times than downloading hitomi.la as a mirror of exhentai.org(max speed only 1MB/s).

I looked into this one, and they were downloaded at the same rate for me. I even stepped through the code to make sure there was nothing going on that there shouldn't be. One thing I did notice, though, is that the speed calculation is off if you resume a download that had previously already been started. I wonder if that's what made it appear slower.
